### PR TITLE
Use `UpgradeCode` for matching MSI apps

### LINKED ISSUE #1145: Add MSI UpgradeCode to manifest
# MSI Installer UpgradeCode

MSI Installers have an UpgradeCode to help with package upgrades:
[Using an UpgradeCode]https://docs.microsoft.com/en-us/windows/win32/msi/using-an-upgradecode)

### PR DESCRIPTION
**Pull request by lechacon**
This change adds support for using an installer's `UpgradeCode` from the `AppsAndFeaturesEntries` for matching against the installed apps.

* Added a new version of the index interface to create a new table for holding the UpgradeCode and a new version of the SearchResultsTable
* Added UpgradeCode as a `PackageVersionMultiProperty` and `SystemReferenceString`
* Updated ARP correlation to also consider UpgradeCode
* Updated the ProductCode and PackageFamilyName tables to also include the data from `AppsAndFeaturesEntries`
* Added extraction of the UpgradeCode for MSI apps when reading the ARP

Closes #1145

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2418)

### CODE REVIEW

File: src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp [hunk 1]
```diff
@@ -3,11 +3,57 @@
...
+        {
+            // The UpgradeCode is not stored in the ARP registry keys, so we use the MSI API to query it.
+            //
+            // The UpgradeCode is apparently also stored in the registry under
+            //   HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UpgradeCodes\<UpgradeCode>
+            // but:
+            //   * This key does not seem to be documented
+            //   * The format used to store the GUIDs is harder to parse
+            //   * We cannot look up by ProductCode; we need to take an UpgradeCode and then find the ProductCode
+            // Using the registry could potentially be faster, so we may consider using it if this proves to be too slow.
```
Comment (lechacon) [inline src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp:25]: This turned out to be quite slow once I actually tested it, so I'll look into other ways of getting the UpgradeCode

File: src/AppInstallerRepositoryCore/Microsoft/Schema/1_1/Interface_1_1.cpp [hunk 2]
```diff
@@ -29,25 +29,36 @@ namespace AppInstaller::Repository::Microsoft::Schema::V1_1
     {
         std::vector<Utility::NormalizedString> GetSystemReferenceStrings(
             const Manifest::Manifest& manifest,
-            std::function<const Utility::NormalizedString&(const Manifest::ManifestInstaller&)> func)
+            std::function<const Utility::NormalizedString&(const Manifest::ManifestInstaller&)> extractStringFromInstaller,
+            std::optional<std::function<const Utility::NormalizedString& (const Manifest::AppsAndFeaturesEntry&)>> extractStringFromAppsAndFeaturesEntry = {})
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Microsoft/Schema/1_1/Interface_1_1.cpp:33]: A `std::function` can already be empty; no need for a `std::optional` wrapper.

File: src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp [hunk 3]
```diff
@@ -3,11 +3,133 @@
...
 
 namespace AppInstaller::Repository::Microsoft
 {
     using namespace AppInstaller::Registry::Portable;
 
+    namespace
+    {
+        // "Unpacks" a GUID in the format used by the UpgradesCode registry key into the usual format.
+        // Returns empty if it is not a valid GUID
+        std::optional<std::string> TryUnpackUpgradeCodeGuid(std::string_view packed)
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp:16]: As discussed, there is a simpler way to do this just using a transformation map from the given 32 character input into the desired GUID-like output.  Up to you on whether to change to that.

I do want to confirm that the output format here does match the value that would be returned by `MsiGetProperty`, as wingetcreate et al. will likely use that function to read the property out of the MSI files themselves when creating manifests.

**Review by JohnMcPMS — APPROVED** [review]

File: src/AppInstallerRepositoryCore/Microsoft/Schema/1_1/Interface_1_1.cpp (hunk 2 above)
Comment (lechacon) [reply src/AppInstallerRepositoryCore/Microsoft/Schema/1_1/Interface_1_1.cpp]: Thanks. Didn't know that

File: src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp (hunk 3 above)
Comment (lechacon) [reply src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp:15]: Changed to use the transformation map.

Yes, it does match.

File: src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp [hunk 4]
```diff
@@ -3,11 +3,133 @@
...
+            std::string_view part2 = packed.substr(Part2Start, Part2Length);
+            std::string_view part3 = packed.substr(Part3Start, Part3Length);
+
+            std::string_view part4[Part4Count];
+            for (size_t i = 0; i < Part4Count; ++i)
+            {
+                part4[i] = packed.substr(Part4Start + i * Part4BlockLength, Part4BlockLength);
+            }
+
+            std::string unpacked;
```
[the remark this answers is not in the document]
Comment (lechacon) [reply src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp]: I'm now initializing to a GUID with just 0s to set the characters that are not mapped.

File: src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp [hunk 5]
```diff
@@ -8,6 +8,102 @@ namespace AppInstaller::Repository::Microsoft
...
+            //
+            // Under 'UpgradeCodes' there is one key for each upgrade code, and each upgrade code key
+            // contains the product code as a value. All the upgrade codes and product codes are GUIDs,
+            // but represented in an unusual way - see TryUnpackUpgradeCodeGuid()
+
+            AICLI_LOG(Repo, Info, << "Reading MSI UpgradeCodes");
+            std::map<std::string, std::string> upgradeCodes;
+
+            // There is no UpgradeCodes key on the x86 view of the registry
+            Registry::Key upgradeCodesKey = Registry::Key::OpenIfExists(HKEY_LOCAL_MACHINE, "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Installer\\UpgradeCodes", 0, KEY_READ | KEY_WOW64_64KEY);
```
Comment (lechacon) [inline src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp:81]: Small important change since the review to fix test failure.

Comment (JohnMcPMS) [reply src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp:81]: It does make sense that these would only be stored in the OS architecture.

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**